lib/rules/lint-template-length.js

```diff
@@ -0,0 +1,93 @@
+'use strict';
+
+/*
+ Enforce template size constraints
+
+ The following values are valid configuration:
+
+   * boolean -- `true` for enabled (same as `unix`) / `false` for disabled
+   * object --
+     * max {number} - the longest a template should be without failing a test (assuming the
+       right thing to do would be to split the template up into pieces.
+     * min {number} - the shortest a template should be without failing a test (assuming the
+       right thing to do would be to inline the template.
+ */
+
+const Rule = require('./base');
+
+const MAX_LENGTH = 100;
+const MIN_LENGTH = 5;
+
+const DEFAULT_CONFIG = {
+  max: MAX_LENGTH,
+  min: MIN_LENGTH,
+};
+
+function isValidConfigObjectFormat(config) {
+  for (let key in config) {
+    let value = config[key];
+    let valueIsInteger = Number.isInteger(value);
+
+    if (key === 'min' && !valueIsInteger) {
+      return false;
+    } else if (key === 'max' && !valueIsInteger) {
+      return false;
+    }
+  }
+  return true;
+}
+
+module.exports = class LengthValidation extends Rule {
+  parseConfig(config) {
+    let configType = typeof config;
+
+    let errorMessage = 'The template-length rule accepts one of the following values.\n ' +
+      '  * boolean - `true` to enable / `false` to disable\n' +
+      '  * object -- An object with the following keys:' +
+      '    * `min` -- Minimum length of a template ' +
+      '    * `max` -- Maximum length of a template. ' +
+      '\nYou specified `' + JSON.stringify(config) + '`';
+
+    switch (configType) {
+    case 'boolean':
+      // if `true` use `DEFAULT_CONFIG`
+      return config ? DEFAULT_CONFIG : {};
+    case 'object':
+      if (isValidConfigObjectFormat(config)) {
+        return config;
+      } else {
+        throw new Error(errorMessage);
+      }
+    case 'undefined':
+      return {};
+    default:
+      throw new Error(errorMessage);
+    }
+  }
+
+  visitor() {
+    return {
+      Program(node) {
+        if (node.loc.start.line <= 1 && node.loc.start.column === 0 && node.loc.end.column === 0) {
```

Comment:
What is this guarding?

Reply:
This is a hack that is no longer necessary when using `this.source`. I seemed that this callback would get called multiple times per template-file, and I wanted to ensure that it was being called for the entire file and not for a snippet. This if guards that we are starting on the first line and column. Also guards that we are ending on the first column on the last line, which only works if every template in the system ends in a newline.

Reply:
Actually, this is still needed, otherwise you get stuff like this:
```
/Users/psbanka/dev/copy-tango/app/pods/onboarding/step1/template.hbs
  5:19  error  Template length of 129 exceeds 100  template-length
  30:86  error  Template length of 129 exceeds 100  template-length
  18:106  error  Template length of 129 exceeds 100  template-length
  41:101  error  Template length of 129 exceeds 100  template-length
  9:30  error  Template length of 129 exceeds 100  template-length
... etc
  -:-  error  Template length of 129 exceeds 100  template-length
```

So I left the `node.loc.start.line <= 1 && node.loc.start.column === 0 ` guard in, but checking for last column == 0 seemed to be unnecessary.

Reply:
FWIW, the reason for this is that every block statement also has a `Program`.  What you are doing with that conditional is roughly ensuring that we only check the _first_ `Program` node.

A simpler (and easier to grok for our future selves) check might be:

```js
visitor() {
  let seenOuterProgram = false;
  return {
    Program: {
      exit(node) {
        if (seenOuterProgram === false) { return; }
        // current logic goes here  
        seenOuterProgram = true;
      }
    }
  }
}
```

Reply:
Nice. Thanks!